File: ice-core/src/main/java/com/ice/core/relation/And.java
package com.ice.core.relation;

import com.ice.common.enums.NodeRunStateEnum;
import com.ice.core.base.BaseNode;
import com.ice.core.base.BaseRelation;
import com.ice.core.context.IceContext;
import com.ice.core.utils.IceLinkedList;

/**
 * @author waitmoon
 * relation AND
 * return false on first false
 * have FALSE--FALSE
 * without FALSE have TRUE--TRUE
 * without children--NONE
 * all NONE--NONE
 */
public final class And extends BaseRelation {

    /*
     * process relation and
     */
    @Override
    protected NodeRunStateEnum processNode(IceContext ctx) {
        IceLinkedList<BaseNode> children = this.getChildren();
        if (children == null || children.isEmpty()) {
            return NodeRunStateEnum.NONE;
        }
        boolean hasTrue = false;
        for (IceLinkedList.Node<BaseNode> listNode = children.getFirst(); listNode != null; listNode = listNode.next) {
            BaseNode node = listNode.item;
            if (node != null) {
                NodeRunStateEnum stateEnum = node.process(ctx);
                if (stateEnum == NodeRunStateEnum.FALSE) {
                    return NodeRunStateEnum.FALSE;
                }
                if (!hasTrue) {
                    hasTrue = stateEnum == NodeRunStateEnum.TRUE;
                }
            }
        }

        if (hasTrue) {
            return NodeRunStateEnum.TRUE;
        }
        return NodeRunStateEnum.NONE;
    }
}

File: sdks/go/relation/and.go
// Package relation provides relation node implementations.
package relation

import (
	stdctx "context"

	icecontext "github.com/zjn-zjn/ice/sdks/go/context"
	"github.com/zjn-zjn/ice/sdks/go/enum"
	"github.com/zjn-zjn/ice/sdks/go/node"
)

// And is a relation node that returns FALSE on the first FALSE.
// - Has FALSE -> FALSE
// - Without FALSE, has TRUE -> TRUE
// - Without children -> NONE
// - All NONE -> NONE
type And struct {
	node.Relation
}

// NewAnd creates a new And relation node.
func NewAnd() *And {
	return &And{
		Relation: *node.NewRelation(),
	}
}

// Process implements the Node interface.
func (a *And) Process(ctx stdctx.Context, roam *icecontext.Roam) enum.RunState {
	return node.ProcessWithBase(ctx, &a.Base, roam, a.processNode, nil)
}

func (a *And) processNode(ctx stdctx.Context, roam *icecontext.Roam) enum.RunState {
	if a.Children == nil || a.Children.IsEmpty() {
		return enum.NONE
	}

	hasTrue := false
	for listNode := a.Children.First(); listNode != nil; listNode = listNode.Next {
		n := listNode.Item
		if n != nil {
			state := n.Process(ctx, roam)
			if state == enum.FALSE {
				return enum.FALSE
			}
			if !hasTrue && state == enum.TRUE {
				hasTrue = true
			}
		}
	}

	if hasTrue {
		return enum.TRUE
	}
	return enum.NONE
}
